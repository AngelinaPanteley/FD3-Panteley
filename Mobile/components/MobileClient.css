.MobileClient
{
  padding: 2px 0;
  display: flex;
  align-items: center;
}

.MobileClientBalance
{
  width: 100px;
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
}

.MobileClientFio
{
  display: inline-flex;
  width: 300px;
  padding-right: 20px;
  justify-content: space-between;
  box-sizing: border-box;
  align-items: center;
}

.MobileClientFioForm {
  display: inline-flex;
}

.MobileClientFioForm input {
  width: 30%;
}

.MobileClientFioContainer {
  display: inline-flex;
  justify-content: space-between;
  width: 100%;
}

.MobileClientActions {
  display: inline-block;
  margin-left: 20px;
}

.MobileClientActions button {
  margin: 0 5px;
}

.MobileClientStatus {
  display: inline-block;
  margin-left: 25px;
  width: 50px;
}

.MobileClientStatusActive {
  color: green;
}

.MobileClientStatusBlocked {
  color: red;
}

.MobileClientId {
  display: inline-block;
  margin-top: -6px;
  margin-right: 10px;
}