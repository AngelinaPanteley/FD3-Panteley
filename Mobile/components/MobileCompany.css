.MobileCompany {
  border: double gray 3px;
  display: inline-block;
  padding: 3px 7px;
  background-color: lightyellow;
}

.MobileCompanyName {
  font-weight: bold;
}

.MobileCompanyClients {
  padding-top: 5px;
}

.MobileNewClientButton {
  margin-top: 20px;
}

.MobileCompanyChangeTitle {
  margin-bottom: 10px;
  display: flex;
  align-items: center;
  justify-content: space-around;
}