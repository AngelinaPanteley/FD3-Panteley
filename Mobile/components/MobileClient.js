import React from 'react';
import PropTypes from 'prop-types';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      fam: PropTypes.string.isRequired,
      im: PropTypes.string.isRequired,
      otch: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
      isEdit: PropTypes.bool.isRequired,
    }),
    onDelete: PropTypes.func.isRequired,
    onFIOChange: PropTypes.func.isRequired,
    onBalanceChange: PropTypes.func.isRequired,
  };

  state = {
    fam: this.props.client.fam,
    im: this.props.client.im,
    otch: this.props.client.otch,
    balance: this.props.client.balance,
    isActive: this.props.client.balance >= 0,
    isEdit: this.props.client.isEdit,
  };

  componentWillReceiveProps = (newProps) => {
    const {fam, im, otch} = this.state;
    const newClient = newProps.client;

    if (fam !== newClient.fam || im !== newClient.im || otch !== newClient.otch) {
      this.setState({
        fam: newClient.fam,
        im: newClient.im,
        otch: newClient.otch,
        isEdit: false,
      });
    }

    if (this.state.balance !== newClient.balance) {
      this.setState({
        balance: newClient.balance,
        isActive: newClient.balance >= 0,
      });
    }

    if (this.state.isEdit !== newClient.isEdit) {
      this.setState({
        isEdit: newClient.isEdit,
      });
    }
  };

  deleteHandle = () => {
    this.props.onDelete(this.props.client.id);
  }

  editButtonHandle = () => {
    this.setState({ isEdit: true });
  }

  increaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, this.state.balance + 1);
  }

  decreaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, this.state.balance - 1);
  }

  saveForm = (e) => {
    e.preventDefault();
    const form = e.target;
    const FIO = {
      fam: form.fam.value,
      im: form.im.value,
      otch: form.otch.value,
    };
    this.props.onFIOChange(this.props.client.id, FIO);
  }

  render() {
    console.log("MobileClient id=" + this.props.client.id + " render");

    const {fam, im, otch} = this.state;

    return (
      <div>
        <div className='MobileClient'>
          <span className="MobileClientId">{this.props.client.id}</span>
          <div className='MobileClientFio'>
            {
              this.state.isEdit
                ?
                <form onSubmit={this.saveForm} className="MobileClientFioForm">
                  <input type="text" defaultValue={fam} name="fam" />
                  <input type="text" defaultValue={im} name="im" />
                  <input type="text" defaultValue={otch} name="otch" />
                  <input type="submit" value="Save" />
                </form>
                :
                <div className="MobileClientFioContainer">
                  <span>{fam + " " + im + " " + otch}</span>
                  <button onClick={this.editButtonHandle}>Edit</button>
                </div>
            }
          </div>
          <div className='MobileClientBalance'>
            <button onClick={this.decreaseBalance}>-</button>
            <span>{this.state.balance}</span>
            <button onClick={this.increaseBalance}>+</button>
          </div>
          <div className='MobileClientStatus'>
            {
              this.state.isActive
                ?
                <span className='MobileClientStatusActive'>Active</span>
                :
                <span className='MobileClientStatusBlocked'>Blocked</span>
            }
          </div>
          <div className='MobileClientActions'>
            <button onClick={this.deleteHandle}>Delete</button>
          </div>
        </div>
      </div>
    );

  }

}

export default MobileClient;
