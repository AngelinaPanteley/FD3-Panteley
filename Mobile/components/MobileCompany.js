import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients.map((c) => {
      c.isShown = true;
      return c;
    }),
  };

  setName = (e) => {
    this.setState({ name: e.target.value });
  };

  setClientBalance = (clientId, newBalance) => {
    let changed = false;
    let newClients = [...this.state.clients];
    newClients.forEach((c, i) => {
      if (c.id == clientId) {
        let newClient = { ...c };
        newClient.balance = newBalance;
        newClients[i] = newClient;
        changed = true;
      }
    });
    if (changed)
      this.setState({ clients: newClients });
  };

  setClientFIO = (clientId, newFIO) => {
    let changed = false;
    let newClients = [...this.state.clients];

    newClients.forEach((c, i) => {
      if (c.id == clientId && (c.fam !== newFIO.fam || c.im !== newFIO.im || c.otch !== newFIO.otch)) {
        let newClient = { ...c };
        newClient.fam = newFIO.fam;
        newClient.im = newFIO.im;
        newClient.otch = newFIO.otch;
        newClients[i] = newClient;
        changed = true;
      }
    });
    if (changed)
      this.setState({ clients: newClients });
  };

  deleteClient = (clientId) => {
    let newClients = [...this.state.clients];
    newClients.forEach((c, i) => {
      if (c.id == clientId) {
        newClients.splice(i, 1);
        return;
      }
    });
    this.setState({ clients: newClients });
  };

  addClient = () => {
    const currentClients = this.state.clients;
    const newClientId = currentClients.length === 0? 1 : currentClients[currentClients.length - 1].id + 1;
    const newClients = [...this.state.clients];
    const newClient = {
      id: newClientId,
      fam: '',
      im: '',
      otch: '',
      balance: this.state.filter === 'blocked'? -1 : 0,
      isEdit: true,
      isShown: true,
    }
    newClients.push(newClient);
    this.setState({ clients: newClients });
  }

  filterClients = (e) => {
    const newFilter = e.target.value;
    if (newFilter !== this.state.filter) {
      let newClients = [...this.state.clients];
      let changed = false;

      const changeFlag = (c, i, flagValue) => {
        const newClient = { ...c };
        newClient.isShown = flagValue;
        newClients[i] = newClient;
        changed = true;
      };

      if (newFilter === 'all') {
        newClients.forEach((c, i) => {
          if (!c.isShown) {
            changeFlag(c, i, true);
          }
        });
      } else if (newFilter === 'active') {
        newClients.forEach((c, i) => {
          if (c.balance >= 0 && !c.isShown) {
            changeFlag(c, i, true);
          } else if (c.balance < 0 && c.isShown) {
            changeFlag(c, i, false);
          }
        });
      } else if (newFilter === 'blocked') {
        newClients.forEach((c, i) => {
          if (c.balance >= 0 && c.isShown) {
            changeFlag(c, i, false);
          } else if (c.balance < 0 && !c.isShown) {
            changeFlag(c, i, true);
          }
        });
      }

      this.setState({
        filter: newFilter,
      });

      if (changed) {
        this.setState({
          clients: newClients,
        });
      }
    }
  }

  render() {

    console.log("MobileCompany render");

    const clients = this.state.clients.map(client => {
      return <MobileClient key={client.id}
        client={client}
        onDelete={this.deleteClient}
        onFIOChange={this.setClientFIO}
        onBalanceChange={this.setClientBalance} />
    });

    return (
      <div className='MobileCompany'>
        <div className="MobileCompanyChangeTitle">
          <div>
            <input type="button" value="МТС" onClick={this.setName} />
            <input type="button" value="Velcom" onClick={this.setName} />
          </div>
          <div className="NameMobileCompanyFilter">
            <label htmlFor="filter">Show</label>
            <select id="filter" value={this.state.filter} onChange={this.filterClients}>
              <option value="all">All</option>
              <option value="active">Active</option>
              <option value="blocked">Blocked</option>
            </select>
          </div>
        </div>
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <div className='MobileCompanyClients'>
          {clients}
        </div>
        <button className="MobileNewClientButton" onClick={this.addClient}>
          Add new client
        </button>
      </div>
    );

  }

}

export default MobileCompany;
