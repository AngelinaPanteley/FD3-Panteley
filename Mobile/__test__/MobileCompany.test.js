"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { expect as chaiExpect } from 'chai';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.1';

import MobileCompany from '../components/MobileCompany';
import MobileClient from '../components/MobileClient.js';

configure({ adapter: new Adapter() });

describe('>>>MobileCompany', () => {

  //SNAPSHOTS

  let companyName = 'Velcom';
  let clientsArr = [
    { id: 101, fam: "Иванов", im: "Иван", otch: "Иванович", balance: 2 },
    { id: 105, fam: "Сидоров", im: "Сидор", otch: "Сидорович", balance: 3 },
    { id: 110, fam: "Петров", im: "Пётр", otch: "Петрович", balance: 1 },
    { id: 120, fam: "Григорьев", im: "Григорий", otch: "Григорьевич", balance: 5 },
  ];

  const companyComponent = <MobileCompany name={companyName} clients={clientsArr} />;

  const component = renderer.create(companyComponent);

  it('renders company correctly', () => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  let companyName2 = 'MTC';
  let clientsArr2 = [];

  const companyComponent2 = <MobileCompany name={companyName2} clients={clientsArr2} />;

  const component2 = renderer.create(companyComponent2);

  it('renders empty array correctly', () => {
    let componentTree = component2.toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  //SHALLOW RENDERING

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(companyComponent);
  });

  it('renders company`s name', () => {
    chaiExpect(wrapper.contains(<div className='MobileCompanyName'>Компания &laquo;{companyName}&raquo;</div>)).to.equal(true);
  });

  it('renders company`s name', () => {
    wrapper.instance().deleteClient(101);
    expect(wrapper.state('clients')[0].id).toEqual(105);
  });

});
