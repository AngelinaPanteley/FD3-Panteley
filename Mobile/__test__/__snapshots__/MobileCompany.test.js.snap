// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>MobileCompany renders company correctly 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyChangeTitle"
  >
    <div>
      <input
        onClick={[Function]}
        type="button"
        value="МТС"
      />
      <input
        onClick={[Function]}
        type="button"
        value="Velcom"
      />
    </div>
    <div
      className="NameMobileCompanyFilter"
    >
      <label
        htmlFor="filter"
      >
        Show
      </label>
      <select
        id="filter"
        onChange={[Function]}
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="active"
        >
          Active
        </option>
        <option
          value="blocked"
        >
          Blocked
        </option>
      </select>
    </div>
  </div>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div
    className="MobileCompanyClients"
  >
    <div
      className="MobileClient"
    >
      <span
        className="MobileClientId"
      >
        101
      </span>
      <div
        className="MobileClientFio"
      >
        <div
          className="MobileClientFioContainer"
        >
          <span>
            Иванов Иван Иванович
          </span>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </div>
      <div
        className="MobileClientBalance"
      >
        <button
          className="MobileClientDecreaseBalance"
          onClick={[Function]}
        >
          -
        </button>
        <span>
          2
        </span>
        <button
          className="MobileClientIncreaseBalance"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="MobileClientStatus"
      >
        <span
          className="MobileClientStatusActive"
        >
          Active
        </span>
      </div>
      <div
        className="MobileClientActions"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <span
        className="MobileClientId"
      >
        105
      </span>
      <div
        className="MobileClientFio"
      >
        <div
          className="MobileClientFioContainer"
        >
          <span>
            Сидоров Сидор Сидорович
          </span>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </div>
      <div
        className="MobileClientBalance"
      >
        <button
          className="MobileClientDecreaseBalance"
          onClick={[Function]}
        >
          -
        </button>
        <span>
          3
        </span>
        <button
          className="MobileClientIncreaseBalance"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="MobileClientStatus"
      >
        <span
          className="MobileClientStatusActive"
        >
          Active
        </span>
      </div>
      <div
        className="MobileClientActions"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <span
        className="MobileClientId"
      >
        110
      </span>
      <div
        className="MobileClientFio"
      >
        <div
          className="MobileClientFioContainer"
        >
          <span>
            Петров Пётр Петрович
          </span>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </div>
      <div
        className="MobileClientBalance"
      >
        <button
          className="MobileClientDecreaseBalance"
          onClick={[Function]}
        >
          -
        </button>
        <span>
          -1
        </span>
        <button
          className="MobileClientIncreaseBalance"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="MobileClientStatus"
      >
        <span
          className="MobileClientStatusBlocked"
        >
          Blocked
        </span>
      </div>
      <div
        className="MobileClientActions"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <span
        className="MobileClientId"
      >
        120
      </span>
      <div
        className="MobileClientFio"
      >
        <div
          className="MobileClientFioContainer"
        >
          <span>
            Григорьев Григорий Григорьевич
          </span>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </div>
      <div
        className="MobileClientBalance"
      >
        <button
          className="MobileClientDecreaseBalance"
          onClick={[Function]}
        >
          -
        </button>
        <span>
          5
        </span>
        <button
          className="MobileClientIncreaseBalance"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <div
        className="MobileClientStatus"
      >
        <span
          className="MobileClientStatusActive"
        >
          Active
        </span>
      </div>
      <div
        className="MobileClientActions"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
  </div>
  <button
    className="MobileNewClientButton"
    onClick={[Function]}
  >
    Add new client
  </button>
</div>
`;

exports[`>>>MobileCompany renders empty array correctly 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyChangeTitle"
  >
    <div>
      <input
        onClick={[Function]}
        type="button"
        value="МТС"
      />
      <input
        onClick={[Function]}
        type="button"
        value="Velcom"
      />
    </div>
    <div
      className="NameMobileCompanyFilter"
    >
      <label
        htmlFor="filter"
      >
        Show
      </label>
      <select
        id="filter"
        onChange={[Function]}
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="active"
        >
          Active
        </option>
        <option
          value="blocked"
        >
          Blocked
        </option>
      </select>
    </div>
  </div>
  <div
    className="MobileCompanyName"
  >
    Компания «
    MTC
    »
  </div>
  <div
    className="MobileCompanyClients"
  />
  <button
    className="MobileNewClientButton"
    onClick={[Function]}
  >
    Add new client
  </button>
</div>
`;
